File: Server/models/materias.go
package models

import (
	"Server/config"
	"context"
	"log"

	"github.com/jackc/pgx/v5"
)

type Materia struct {
	Codigo        string    `json:"codigo"`
	Nombre        string    `json:"nombre"`
	Info          string    `json:"info"`
	Uc            int       `json:"uc"`
	Horas_estudio string    `json:"horas_estudio"`
	Electiva      bool      `json:"electiva"`
	Correquisito  bool      `json:"correquisito"`
	Id_carrera    int       `json:"id_carrera"`
	Departamento  string    `json:"departamento"`
	Nucleo        string    `json:"nucleo"`
	Semestre      int       `json:"semestre"`
	UC_requeridas *int      `json:"uc_requeridas"`
	Prelaciones   []*string `json:"prelaciones"`
	Debloqueables []*string `json:"desbloqueables"`
}

func GetAllMaterias() ([]Materia, error) {
	query := `
	SELECT 
		m.codigo, 
		m.nombre, 
		m.info, 
		m.uc, 
		m.horas_estudio,
		m.electiva,
		m.correquisito,
		d.nombre,
		n.nombre,
		s.semestre,
		uc.min_uc,
		ARRAY_AGG(DISTINCT prl.codigo_prel) AS prelaciones,
		ARRAY_AGG(DISTINCT des.codigo_mat) AS debloqueables
	FROM materia m
	JOIN departamento d ON m.id_departamento = d.id
	JOIN nucleo n ON m.id_nucleo = n.id
	JOIN semestre_mat_carrera s ON m.codigo = s.codigo_materia
	LEFT JOIN prelacion_uc uc ON m.codigo = uc.codigo_mat
	LEFT JOIN prelacion_mat prl ON m.codigo = prl.codigo_mat
	LEFT JOIN prelacion_mat des ON m.codigo = des.codigo_prel
	GROUP BY s.semestre, m.codigo, d.nombre, n.nombre, uc.min_uc;`

	rows, err := config.PsqlDB.Query(context.Background(), query)
	if err != nil {
		log.Printf("\n\nError getting Materias: %v", err)
		return nil, err
	}
	defer rows.Close()

	var materias []Materia
	for rows.Next() {
		var m Materia
		err := rows.Scan(&m.Codigo, &m.Nombre, &m.Info, &m.Uc, &m.Horas_estudio, &m.Electiva, &m.Correquisito, &m.Departamento, &m.Nucleo, &m.Semestre, &m.UC_requeridas, &m.Prelaciones, &m.Debloqueables)
		if err != nil {
			log.Printf("Error fetching Materias: %v", err)
			return materias, err
		}
		materias = append(materias, m)
	}
	return materias, nil

}

func (m *Materia) GetAllMateriasByCarrera(db *pgx.Conn) ([]Materia, error) {
	query := `
	SELECT 
		m.codigo, 
		m.nombre, 
		m.info, 
		m.uc, 
		m.horas_estudio,
		m.electiva,
		m.correquisito,
		d.nombre,
		n.nombre,
		s.semestre,
		uc.min_uc,
		ARRAY_AGG(DISTINCT prl.codigo_prel) AS prelaciones,
		ARRAY_AGG(DISTINCT des.codigo_mat) AS debloqueables
	FROM materia m
	JOIN departamento d ON m.id_departamento = d.id
	JOIN nucleo n ON m.id_nucleo = n.id
	JOIN semestre_mat_carrera s ON m.codigo = s.codigo_materia
	LEFT JOIN prelacion_uc uc ON m.codigo = uc.codigo_mat
	LEFT JOIN prelacion_mat prl ON s.codigo_materia = prl.codigo_mat AND prl.id_carrera = @id_carrera
	LEFT JOIN prelacion_mat des ON s.codigo_materia = des.codigo_prel AND des.id_carrera = @id_carrera
	WHERE s.id_carrera = @id_carrera
	GROUP BY s.semestre, m.codigo, d.nombre, n.nombre, uc.min_uc;`

	rows, err := config.PsqlDB.Query(context.Background(), query, pgx.NamedArgs{"id_carrera": m.Id_carrera})
	if err != nil {
		log.Printf("\n\nError getting Materias: %v", err)
		return nil, err
	}
	defer rows.Close()

	var materias []Materia
	for rows.Next() {
		var m Materia
		err := rows.Scan(&m.Codigo, &m.Nombre, &m.Info, &m.Uc, &m.Horas_estudio, &m.Electiva, &m.Correquisito, &m.Departamento, &m.Nucleo, &m.Semestre, &m.UC_requeridas, &m.Prelaciones, &m.Debloqueables)
		if err != nil {
			log.Printf("Error fetching Materias: %v", err)
			return materias, err
		}
		materias = append(materias, m)
	}
	return materias, nil

}

func (m *Materia) GetMateriasDeDepartamento(db *pgx.Conn) (*[]Materia, error) {
	query := `
	SELECT 
		m.codigo, 
		m.nombre, 
		m.info, 
		m.uc, 
		m.horas_estudio,
		m.electiva,
		d.nombre,
		n.nombre,
		s.semestre,
		uc.min_uc,
		ARRAY_AGG(prl.codigo_prel) AS prelaciones
	FROM materia m
	JOIN departamento d ON m.id_departamento = d.id
	JOIN nucleo n ON m.id_nucleo = n.id
	JOIN semestre_mat_carrera s ON m.codigo = s.codigo_materia
	LEFT JOIN prelacion_uc uc ON m.codigo = uc.codigo_mat
	LEFT JOIN prelacion_mat prl ON m.codigo = prl.codigo_mat
	WHERE LIKE d.name = @id_departamento 
	GROUP BY s.semestre, m.codigo, d.nombre, n.nombre, uc.min_uc;`

	rows, err := config.PsqlDB.Query(context.Background(), query, pgx.NamedArgs{"id_departamento": m.Departamento})
	if err != nil {
		log.Printf("\n\nError getting Materia: %v", err)
		return nil, err
	}
	defer rows.Close()

	var materias []Materia
	for rows.Next() {
		var m Materia
		err := rows.Scan(&m.Codigo, &m.Nombre, &m.Info, &m.Uc, &m.Horas_estudio, &m.Electiva, &m.Departamento, &m.Nucleo, &m.Semestre, &m.UC_requeridas, &m.Prelaciones, &m.Debloqueables)
		if err != nil {
			log.Printf("Error fetching Materias: %v", err)
			return &materias, err
		}
		materias = append(materias, m)
	}
	return &materias, nil

}

File: Client/src/actions/materias.ts
import { defineAction } from 'astro:actions';
import { z } from 'astro:content';
import { getSecret } from 'astro:env/server';
import type { IMaterias, IMateriasBySemester, IMateriasObject } from '../schemas/materias';
import { de } from 'zod/v4/locales';

const SERVER_ADDRESS = getSecret("SERVER_ADDRESS")

export const materias = {
    get: defineAction({
        input: z.object({
            name: z.string(),
        }),
        handler: async (input,context) => {
            try {
              const res = await fetch(`http://${SERVER_ADDRESS}/materias/departamento/${name}`,{
                  method: "GET",
                  headers: context.request.headers
              });

              if (!res.ok) {
                  throw new Error(`Error fetching thesis data: ${res.statusText}`);
              }
              return res.json();
            
          } catch (error) {
              throw new Error(`Error fetching thesis data: ${error}`);
          }
        }
    }),
    getAllMateriasByCarrera : defineAction({
        input: z.object({
            id: z.string(),
        }),
        handler: async (input,context) => {
            try {
              const res = await fetch(`http://${SERVER_ADDRESS}/materias/${input.id}/all`,{
                  method: "GET",
                  headers: context.request.headers
              });

              if (!res.ok) {
                  throw new Error(`Error fetching thesis data: ${res.statusText}`);
              }
              return res.json();
            
          } catch (error) {
              throw new Error(`Error fetching thesis data: ${error}`);
          }
        }
    }),
    getAll: defineAction({
       handler: async (input, context) => {
            const res = await fetch(`http://${SERVER_ADDRESS}/materias/all`, {
                method: "GET",
                headers: context.request.headers
            });
            
            if (!res.ok) {
                throw new Error(`Error fetching thesis data: ${res.statusText}`);
            }

            const materias = await res.json();

            const materiasObj = Object.assign({}, ...(materias as IMaterias[]).map(materia=> ({
                [materia.codigo]: {
                    nombre: materia.nombre,
                    info: materia.info,
                    uc: materia.uc,
                    horas: materia.horas_estudio,
                    electiva: materia.electiva,
                    nucleo: materia.nucleo,
                    departamento: materia.departamento,
                    semestre: materia.semestre,
                    uc_requeridas: (materia.uc_requeridas) ? materia.uc_requeridas : 0,
                    prelaciones: materia.prelaciones,
                    desbloqueables: materia.desbloqueables
                }
                })));

            // const resData = await res.json() as IMaterias[];

            // let data = {} as IMateriasBySemester

            // for (const materia of resData ) {
            //     if (!data[materia.semestre]) {
            //         data[materia.semestre] = [];
            //     }
            //    data[materia.semestre].push(materia);
            // }


            return materiasObj as IMateriasObject;
        }
    })
}